### PR TITLE
Conditional login through identity provider

### LINKED ISSUE #20191: Conditional login through identity provider
### Description

Original issue in Jira: [KEYCLOAK-19956](https://issues.redhat.com/browse/KEYCLOAK-19956)

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by dmartinol**
Closes #20191

Added new fields in the OIDC settings page to specify an additional authentication check based on token claims:
- one switch to toggle the check on and off: defaults to off
- one field to define the token name (must be in the ID token section): defaults to empty string
- one field to define the expected token value (with regex matching): defaults to empty string
Latest two fields are hidden if the switch is turned off.

Default values are provided in the IdentityProviderModel, which should guarantee proper upgrade of data from previous release.

I have found no existing unit test on this area, so I did not include a new one for this feature. 

Updated Server Admin guide to reflect the change.

Tested on local environment and also on OpenShift deployment.

### CODE REVIEW

**Comment by dmartinol** [comment]
> This is to close [KEYCLOAK-19956](https://issues.redhat.com/browse/KEYCLOAK-19956)

Created #20191

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:107]: Don't use "private" fields. You should replace this by a "watch"
```suggestion
  const filteredByClaim = useWatch({ control, name: "config.filteredByClaim", defaultValue: false});
```

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:155]: remove debug

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:189]:
```suggestion
              {...register("config.claimFilterName", { required: true })}
```

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:184]: because `claimFilterRequired` is always true this can be shorted:
```suggestion
              isRequired
```

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:183]: also add the validated property here so the field turns red when there is an error

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:185]: no need to add this as `false` is the default and `claimFilterRequired` is always true

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:186]: same here `type="text"` is the default no need to add it

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (edewit) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:217]: this field has similar problems

**Review by edewit — CHANGES_REQUESTED** [review]
I only looked at the frontend code, don't know about this feature

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (dmartinol) [reply js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx]: thanks

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (dmartinol) [reply js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx]: thanks

**Review by edewit — DISMISSED** [review]
:+1:

**Comment by stianst** [comment]
@pedroigor @mposolda could you take a look at this please?

**Comment by stianst** [comment]
Had a quick look at this and couple notes from me:

* General approach looks sane to me, but not sure about the names/descriptions of the fields. Filter by claim is not directly clear to me that it means require claim to be present. Perhaps it could be renamed to something like `Require claim to authenticate` and `Required claim name` / `Required claim value`?
* We need some tests for this, not sure how we can do that, but I imagine we could extend on testing with IdP brokering with another realm as that would be easier to configure/setup

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (mposolda) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:139]: The feature is implemented just for OIDC Identity providers. So are we showing the relevant options just for OIDC Identity providers? As I think it should be hidden for SAML IDP and also for Social Identity providers, which are not OIDC Identity provider.

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:396]: Is it possible to use `tracef` instead of `infof` (in this place and also in all other in this class) and also use `debugf` instead of `warnf` (in this place and also in all other in this class). As AFAIK `info` and `warn` are usually for special messages or server-wide messages/issues, but not for regular "user" requests.

**Review by mposolda — CHANGES_REQUESTED** [review]
I've added few additional inline comments in addition to @stianst comment.

I am maring this as "missing tests". We have identity brokering tests in the testsuite in the package `org.keycloak.testsuite.broker` . Hope you can use some existing tests for the inspiration.

**Review by pedroigor — COMMENTED** [review]
This looks somewhat related to https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter. 

However, the `claims` parameter is `OPTIONAL` and I'm not sure if OPs are supposed to *not* authenticate a user if a requested claim (and optionally its value) is not available for a given user.

I think it makes sense to at least make it more aligned with the idea of the `claims` parameter from the specification. Something like:

* Verify essential claim (ON/OFF. If enabled, ID tokens issued by the OP must have a specific claim. Otherwise, the user can not authenticate through this broker.)
* Essential claim (name)
* Essential claim value (value)

Even though not strictly related to `claims` support, using a similar terminology should help to understand what it does.

File: js/apps/admin-ui/public/locales/en/identity-providers-help.json
Comment (dmartinol) [reply js/apps/admin-ui/public/locales/en/identity-providers-help.json:41]: will do, even if it might change if we also include Stian's comment above:
> Perhaps it could be renamed to something like Require claim to authenticate and Required claim name / Required claim value?

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (dmartinol) [reply js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:139]: Good catch, fixing and revalidating it

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
Comment (dmartinol) [reply services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:396]: Thanks for sharing these practices

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
Comment (dmartinol) [reply services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:398]: No special reasons. I will update the code and the documentation as well

**Comment by dmartinol** [comment]
> I am marking this as "missing tests". We have identity brokering tests in the testsuite in the package `org.keycloak.testsuite.broker` . Hope you can use some existing tests for the inspiration.

@mposolda while writing the test code I verified that `UserRepresentation.setAttributes` method has a `Map<String, List<String>>`, as if for each user attribute we could define multiple values, which I think it's not doable from the Keycloak Admin console. As a result, the OIDC provider receives an ID token claim that is a JSON array and cannot easily compare it with the filter configured in the OIDC configuration.
I can change the provider to look at the actual type of claim value, in this case a `java.util.List`, and apply the match on any of the contained values, but I'd also like to understand why I saw this behavior only with the test code while the debugging on the development server only returned claim values of `String` type.

**Comment by dmartinol** [comment]
@pedroigor @mposolda integrated requested changed (apart from labels for which I'm waiting for some final decision)

**Comment by pedroigor** [comment]
@dmartinol I think the remaining tasks are basically around the names you are using to configure the essential|required claims.

I would use something like what I proposed (and it seems @mposolda agrees about it too). It is also aligned with the request from @stianst to make it more meaningful.

**Comment by dmartinol** [comment]
> @dmartinol I think the remaining tasks are basically around the names you are using to configure the essential|required claims.
> 
> I would use something like what I proposed (and it seems @mposolda agrees about it too). It is also aligned with the request from @stianst to make it more meaningful.

@mposolda @pedroigor @stianst I believe I have submitted all required changes including integration testing, could you please clarify if anything else is missing or if there are other concerns to move forward with this PR?

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:153]: Instead of declaring a new variable and concatenating a string to do a cast, you could call [`toString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean/toString) instead.
```suggestion
                    field.onChange(value.toString())
```

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:157]: An `aria-label` attribute should have a human readable value.

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:220]: I am not sure you can update the `required` value here after it has been passed. It might have to be set explicitly with a different API.

Have you tested that this works as expected?

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by jonkoops** [comment]
@dmartinol there are several linting errors in the front-end code, could you fix these?

**Comment by dmartinol** [comment]
> @dmartinol there are several linting errors in the front-end code, could you fix these?

sure, I will give it a try, thanks for reviewing the changes

**Comment by dmartinol** [comment]
@jonkoops where can I see the `linting errors in the front-end code`?
Submitted the other changes in the meantime

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:158]: This can be removed as the field is already labeled through `FormGroupField`.

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:110]: Not sure, but likely this should be a string to match the actual value stored in the form.
```suggestion
    defaultValue: "false",
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (dmartinol) [reply js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:158]: Removing "aria-label" field thanks

File: js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx
Comment (dmartinol) [reply js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx]: Not sure why it worked before but you are correct, it's a string

**Review by jonkoops — DISMISSED** [review]
Changes from the front-end side of things are looking good to me.

**Comment by dmartinol** [comment]
@mposolda looks like you requested one more change but, apart from the test code that was already submitted, I can't see any other requests, could you please verify if I'm missing anything?

**Comment by mposolda** [comment]
@dmartinol There are failing tests and looks that they are regression of this PR. I suggest to rebase and fix the failing tests. IMO this PR is good to go once that is done.

**Comment by dmartinol** [comment]
> Details

@mposolda now seems that some "Admin UI E2E" tests are failing, do I need to install cypress and fix the broken test `identity_providers_test.spec.ts`?

**Comment by jonkoops** [comment]
@dmartinol yes, if possible please also fix the tests. You can find instructions on how to run the tests in [the README](https://github.com/keycloak/keycloak/tree/main/js/apps/admin-ui#integration-testing-with-cypress).

**Comment by dmartinol** [comment]
> @dmartinol There are failing tests and looks that they are regression of this PR. I suggest to rebase and fix the failing tests. IMO this PR is good to go once that is done.

Fixed broken tests: the page became longer and the initial switched in the advanced settings page were not visible, so I added a click on the `Advanced Settings` link to restore the page.
Also added check that the new switch and text fields are populated

**Review by jonkoops — CHANGES_REQUESTED** [review]
There are some linting issues in the front-end that need to be resolved.

**Comment by dmartinol** [comment]
> There are some linting issues in the front-end that need to be resolved.

It took me 38 commits, but fInally "All checks have passed"

**Comment by trixpan** [comment]
why not using "Post login flow" for achieving the same goal?

**Comment by trixpan** [comment]
But isn't the use creation controlled by First Login Flow? 

That seems like a better fit than creating an option that will be tightly linked to the IdP section of the login flow.

**Review by jonkoops — DISMISSED** [review]
Changed to the front-end LGTM

**Review by andymunro — APPROVED** [review]
Looks good, @jonkoops

**Comment by trixpan** [comment]
> you mean to develop a new Condition to validate the required claim and put it before the "Create User if Unique" to exit the flow in case the claim is not matched?

`Create User if Unique` is not a requirement of that flow, many deployments will simply link to existing users using `Detect Existing Broker User` followed by auto linking as documented. 

I would say that a conditional, added to the flow, would be far more natural than a filter linked to a particular IdP.

Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

**Comment by bmozaffa** [comment]
> Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

Definitely needs to be verified, but at least my impression was that the user is created locally before IdP claims are imported as user attributes, so this would be a chicken and egg problem if we're trying to avoid initial login/user creation

**Comment by dmartinol** [comment]
> Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

If we use that condition it means that a) we created an IDP mapper to convert the claim to the attribute and b) (I guess) the user has to be created together with attributes
So this would not solve the request to not create the user

**Comment by trixpan** [comment]
> > Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?
> 
> If we use that condition it means that a) we created an IDP mapper to convert the claim to the attribute and b) (I guess) the user has to be created together with attributes So this would not solve the request to not create the user

If that is the case, again, haven't tested, you don't need to use that particular step provider. Aby other purpose built execution provider would be perfectly fine. 

The point I am trying to make is: there is a flow for post authentication, that flow clearly has steps that happen before the user is created (as you may chose to use the existing user create a new one). Sticking to that pattern of flow based checks seems cleaner that bolting something into the IdP implementation.

In the end, the result is the same. One (via flow) will work will all providers, the other one will depend on class inheritance or re-implementation. But access will be conditional to attributes.

**Comment by pedroigor** [comment]
@trixpan I share the same but we can't guarantee that all protocols can share the same mechanism for handling user attributes. It should work if we enable it in the `IdpCreateUserIfUniqueAuthenticator` authenticator, for both OIDC and SAML. There the user is still transient and we can access the claims/attributes.

For now, I would keep it specific to OIDC and enable support for SAML if there are demands for it.

**Review by pedroigor — DISMISSED** [review]

**Comment by dmartinol** [comment]
@mposolda seems that there are more changes required from you but cannot find anything. could you please verify and let me know what else is missing?

**Comment by stianst** [comment]
Would be worth adding this to release notes (https://github.com/keycloak/keycloak/blob/main/docs/documentation/release_notes/topics/22_0_0.adoc), as I know this has been raised a few times at least in the past.

**Comment by mposolda** [comment]
@dmartinol LGTM, Thanks for the tests! As long as it is added to release notes as proposed by @stianst , we should be good with this PR from my point of view.

@pedroigor @trixpan @dmartinol Handling this at the first-broker-login-flow/post-broker-login-flow may be also possible, but I am not sure if it's better than adding the config options directly into the Idp as done in this PR. Both approaches have advantages and disadvantages though... IMO flow approach might be conceptually cleaner and is independent on protocol, however it can result in many flows created (in case you have many Identity providers and every of them has slightly different set of "required claims").

**Review by mposolda — DISMISSED** [review]
@dmartinol Thanks! LGTM. As long as tests are OK, I will need to squash the commits...

**Review by jonkoops — DISMISSED** [review]

**Review by jonkoops — DISMISSED** [review]

File: docs/documentation/release_notes/topics/22_0_0.adoc
Comment (andymunro) [inline docs/documentation/release_notes/topics/22_0_0.adoc:135]:
```suggestion
The option is disabled by default; when it is enabled, you can specify the name of the JWT token claim to filter and the value to match 
```

File: docs/documentation/server_admin/topics/identity-broker/configuration.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/identity-broker/configuration.adoc:69]:
```suggestion
|When *ON*, ID tokens issued by the identity provider must have a specific claim. Otherwise, the user can not authenticate through this broker

|Essential claim
|When *Verify essential claim* is *ON*, the name of the JWT token claim to filter (match is case sensitive)

|Essential claim value
|When *Verify essential claim* is *ON*, the value of the JWT token claim to match (supports regular expression format)
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good, @jonkoops. A few minor changes to suggest.

**Review by jonkoops — APPROVED** [review]

**Comment by dmartinol** [comment]
> @jonkoops @dmartionol Thanks for the updates, but I am not sure if test failures are just "instability" or if it's regression of this PR? Seems to me that issues in `Admin UI E2E (3, Chrome)` are regression of this PR (Seeing 7 failures in `identity_providers_test.spec.ts` in this group). There is also something in `Admin UI E2E (1, Chrome)`, but looks only as single test failure, so might be instability? Could you confirm?

Now `users_test.spec.ts` is failiing, and I don't think this is affected by the PR, is it? Anyway, I also see random failures on my local env, so I'm not sure what's going on.

**Comment by mposolda** [comment]
@dmartinol I think that those test failures are not caused by your PR, but rather some general instability as I see them also in some other PR https://github.com/keycloak/keycloak/pull/20621 (looks like exactly same test failures)... @jonkoops Could you please confirm?

BTV. It can be helpful if you stop merging this PR with latest Keycloak main (without any changes in your PR) as it just re-triggers the build when the links to the old build would be lost. We can eventually re-trigger your jobs when needed and/or will ask you for rebase/merge.

**Comment by mposolda** [comment]
@dmartinol I've updated the branch to latest Keycloak main. That should probably help to fix test failures (They were fixed in the meantime in Keycloak main by PR https://github.com/keycloak/keycloak/pull/21312 ). Hopefully we're finally very close to merge now :-)

**Review by mposolda — APPROVED** [review]

**Pull request merged**
